Add deprecation info and warnings for an empty TIER_PREFERENCE

joegallo opened the pull request:
Part of https://github.com/elastic/elasticsearch/issues/76147

On 7.x, for clusters that have data nodes that don't have all data roles (e.g. a cluster with a dedicated hot tier would mean those nodes had `data_hot` but not all the other `data_...` roles), we want to emit deprecation warnings when an index is created that doesn't have a tier preference, and raise any such previously created indices with deprecation info.

~Note: I've got the logic mostly nailed down here, or so I think, and there are some reasonable tests as well. Clearly the strings for the warnings and info are placeholders -- I'll take another swing at them on Monday. For everything else, though, feel free to ignore the 'WIP' tag to the right, this is intended to be mostly reviewable.~

joegallo committed:
Organize imports

joegallo committed:
Add a utility method data nodes that are missing roles

joegallo committed:
Thread the ClusterState into IndexDeprecationChecks

so that it's available to use for a check that wants to use it

joegallo committed:
New index deprecation check for empty tier preference

Strings and other details obviously placeholder

joegallo committed:
New deprecation log for empty tier preference

Strings and other details obviously placeholder

joegallo committed:
Add a tier preference for these tests

They don't like the new warning, but we can shortcut it by adding a
tier preference.

henningandersen reviewed (commented):
I left some comments, otherwise this looks good.

joegallo committed:
INDICES is fine

joegallo committed:
Put helpers after tests, rather than before

joegallo committed:
This helper can be private

joegallo committed:
Randomize shard and replica counts

joegallo committed:
Add an all-but-one data roles element to this test

and rewrite the set matching to use hamcrest more heavily -- it was
easy enough to handle a set of one item manually with an iterator, but
with two items in the picture it's a bit annoying to do 'by hand'

joegallo committed:
Merge branch '7.x' into enforce-default-tier-preference-checks

joegallo committed:
Oops, unused import

joegallo committed:
This can just be BiFunction

We don't need a purpose-built interface for these.

joegallo committed:
Merge branch '7.x' into enforce-default-tier-preference-checks

joegallo committed:
Collapse test into testEnforceDefaultTierPreference

If the aggregated settings end up not containing a tier preference,
and the cluster has a node without all the data roles, then we should
emit a warning.

For convenience, the test randomly chooses between an empty cluster
state and one that doesn't have all the data roles, so the final
condition is written in terms of merely not being the empty
state (since the alternative is always one that doesn't have all the
data roles, at least in the universe of this test).

joegallo committed:
Merge branch '7.x' into enforce-default-tier-preference-checks

joegallo committed:
Non-final (but also non-placeholder) messages

We can iterate on these until we get to something that's just right,
but this at least gives us a start.

joegallo committed:
Merge branch '7.x' into enforce-default-tier-preference-checks

masseyke reviewed (approved):
Looks good to me, with the caveat that we need to make sure to return to the deprecation info URL and message at some point soon.

elasticmachine committed:
Merge branch '7.x' into enforce-default-tier-preference-checks
